### PR TITLE
feat(compiler): bypass throw at parent level and disregard return type

### PR DESCRIPTION
**Pull request by Warkanlock**
Based on #5414 

## Description

Added a bypass check at the return type level when a throw is detected in the init statement. In addition, I implemented a more generic visitor iterator that allows us to visit and collect not only the return type of the throw type and potentially more statements. 

Before:
```
error: A function whose return type is "num" must return a value.
  --> wing/main.w:1:24
  |  
1 |   let x: (): num = () => {
  | /------------------------^
2 | |   throw "not implemented";
3 | | };
  | \-^ A function whose return type is "num" must return a value
```

Now this code executed well and it has also tests to support this claim

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: examples/tests/valid/bypass_return.test.w
Comment (MarkMcCulloh) [inline examples/tests/valid/bypass_return.test.w:28]: These tests are at the type-checker level, so no need to use tests.
Also, the last test isn't really different from the second one since we don't do any sort of data flow analysis.

```suggestion
let simpleThrow: (): num = () => {
  throw "not implemented";
};

let throwWithConditional: (): num = () => {
  // this should not be executed
  if false {
    return 1;
  }

  throw "not implemented";
};
```

File: examples/tests/invalid/bypass_return.test.w
Comment (MarkMcCulloh) [inline examples/tests/invalid/bypass_return.test.w:1]: I think some additional cases could be added here. One I can think if so to verify that we still ensure if you (optionally) return the correct type even if a throw happens at the top level


```wing
let wrongReturnInScope = (): num => {
  if false {
    return "a";
         //^^^ Expected type to be "num", but got "str" instead
  }

  throw "error";
}
```

or, if the throw happens in a lower scope we still require a return

```wing
let missingReturn = (): num => {
  if true {
    throw "error";
  }
}
//^ Expected to return "num"
```

**Review by MarkMcCulloh — CHANGES_REQUESTED** [review]
Haven't had a chance to look at the implementation yet, just had some notes on the tests.

btw when writing tests in wing, use 2 spaces for indentation. Sadly we don't have any autoformatting for that yet 😢

File: examples/tests/invalid/bypass_return.test.w
Comment (Warkanlock) [reply examples/tests/invalid/bypass_return.test.w:1]: Thanks, Mark, for the suggestions; I most definitely need to change my Vim config to have 2 tabs instead of 4 whenever I spend time working on wing files.

Are there any plans for the linter of the wing? We might be able to do it pretty fast if we already have the grammar in place

In regards to the testing suite, good to point out that there's no need to test inside a test for a type-checker validation (I thought it was a good practice inside Wing to use it always, my bad)

Comment (MarkMcCulloh) [reply examples/tests/invalid/bypass_return.test.w:1]: > Are there any plans for the linter of the wing?

Yeah absolutely! The goal is for the wing toolchain to be very batteries-included, so linter/formatter would be included (eventually)

> I thought it was a good practice inside Wing to use it always, my bad

In normal wing code that's definitely true, but this corpus of wing files is a little special since it's specifically written as test cases for the compiler

File: examples/tests/invalid/bypass_return.test.w
Comment (Chriscbr) [inline examples/tests/invalid/bypass_return.test.w:20]: Can we add a test case where the function body is empty? I believe it should still raise an error in this scenario

File: libs/wingc/src/type_check.rs
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:3320]: I think the code for `HasReturnStatementVisitor` can be removed

**Review by Chriscbr — COMMENTED** [review]
Looks promising!

File: libs/wingc/src/type_check.rs
Comment (Warkanlock) [reply libs/wingc/src/type_check.rs:3320]: Sure! I'll do, I didn't want to mess with old code but if I have this confirmation I'll proceed

File: examples/tests/invalid/bypass_return.test.w
Comment (Warkanlock) [reply examples/tests/invalid/bypass_return.test.w:1]: Excellent; let me know if there's anything I can help with in the linter!

Comment (MarkMcCulloh) [inline examples/tests/invalid/bypass_return.test.w:1]: In CI this file is compiling successfully when it should have errors

**Review by Chriscbr — APPROVED** [review]

**Comment by staycoolcall911** [comment]
Hey @MarkMcCulloh - do you still want to give it another review before this goes in? Your changes-requested are currently blocking this from merging, so just want to understand if there is still anything missing in this PR. Thanks!

**Review by MarkMcCulloh — APPROVED** [review]

**Pull request merged**
